transport/outbound.go

```diff
@@ -22,7 +22,19 @@ package transport
 
 import "golang.org/x/net/context"
 
-//go:generate mockgen -destination=transporttest/outbound.go -package=transporttest go.uber.org/yarpc/transport Outbound
+//go:generate mockgen -destination=transporttest/outbound.go -package=transporttest go.uber.org/yarpc/transport Outbound,OutboundCall,RequestSender
+
+// RequestSender sends the given request to the remote.
+type RequestSender interface {
+	Send(ctx context.Context, req *Request) (*Response, error)
+}
+
+// OutboundCall is a single outbound call.
+type OutboundCall interface {
+	// WithRequest builds the transport request based on the given transport
+	// options and sends it through the given RequestSender.
+	WithRequest(ctx context.Context, opts Options, sender RequestSender) (*Response, error)
```

Comment:
The intention was to use a pattern analogous to context managers in Python:

    WithResource(..., f)
    
Is basically,

    resource := acquire(...)
    defer resource.free()
    f(resource)

`resource` is valid only inside the `f`. We have used this pattern elsewhere,
I think. That said, I agree that `WithRequest` is perhaps not the best name
for this. I'm happy to name this something else. `SendWithOptions` or similar
does sound nicer.
    
A note about naming besides that: The fact that there's a stack of these is a
detail not exposed in these interfaces. You could use these interfaces with
single objects, not stacks. So naming anything with stack in the name is
undesirable.